core/ExecutePipe.cpp

```diff
@@ -180,43 +206,108 @@ namespace olympia
         if (ready_queue_.size() > 0) {
             issue_inst_.schedule(sparta::Clock::Cycle(0));
         }
+
+        // Schedule completion
+        complete_inst_.preparePayload(ex_inst)->schedule(1);
+
+    }
+
+    // Called by the scheduler, scheduled by complete_inst_.
+    void ExecutePipe::completeInst_(const InstPtr & ex_inst)
+    {
+        ex_inst->setStatus(Inst::Status::COMPLETED);
+        ILOG("Completing inst: " << ex_inst);
     }
 
     void ExecutePipe::flushInst_(const FlushManager::FlushingCriteria & criteria)
     {
         ILOG("Got flush for criteria: " << criteria);
 
-        // Flush instructions in the ready queue
-        ReadyQueue::iterator it = ready_queue_.begin();
+        // Flush instructions in the issue queue
         uint32_t credits_to_send = 0;
-        while(it != ready_queue_.end()) {
-            if((*it)->getUniqueID() >= uint64_t(criteria)) {
-                ready_queue_.erase(it++);
+
+        auto issue_queue_iter = issue_queue_.begin();
+        while (issue_queue_iter != issue_queue_.end())
+        {
+            auto delete_iter = issue_queue_iter++;
+
+            auto inst_ptr = *delete_iter;
+
+            // Remove flushed instruction from issue queue and clear scoreboard callbacks
+            if (criteria.includedInFlush(inst_ptr))
+            {
+                issue_queue_.erase(delete_iter);
+
+                scoreboard_views_[reg_file_]->clearCallbacks(inst_ptr->getUniqueID());
+
                 ++credits_to_send;
-            }
-            else {
-                ++it;
+
+                ILOG("Flush Instruction ID: " << inst_ptr->getUniqueID() << " from issue queue");
             }
```

Comment:
Is the issue queue age ordered? If so, can't we just break when we find the first instruction that is not included in the flush?

Reply:
It should be, I avoided inferring the ordering though. I figured it's worth the extra cycles to make flushing more flexible.  i.e. if we wanted to implement multi-threading.

Reply:
Ah! I didn't think of that. Makes sense.